### ISSUE TITLE
Integration with LightGBM?

### ISSUE THREAD

**Issue by noklam** [allegroai/clearml#199]
Is there a future support for LightGBM? 
![image](https://user-images.githubusercontent.com/18221871/91710349-b0bd6980-ebb6-11ea-91fe-28b88f5025d6.png)

**Comment by bmartinn** [comment]
Hi @noklam ,
That sounds like a great idea! 
A few followup questions:
1. How do you usually store the trained model with lightgbm, [lightgbm.save_model](https://lightgbm.readthedocs.io/en/latest/pythonapi/lightgbm.Booster.html?highlight=save_model#lightgbm.Booster.save_model) ? 

2. How do you load a model from file (I could not find a specific function for that) ?

3. Plotting seems to rely on matplotlib (based on the [docs](https://lightgbm.readthedocs.io/en/latest/Python-API.html#plotting)), this means the there is no need to add another layer of integration as we already auto-magically connect with matplotlib, could you verify ?

**Comment by noklam** [comment]
@bmartinn 
I know two ways of doing so. 
1. using lightgbm library save: lgb.Boost.save_model,  load: `lgb.Booster(model_file=model)`
2. pickling with joblib
```
import joblib
# save model
joblib.dump(my_model, 'lgb.pkl')
# load model
gbm_pickle = joblib.load('lgb.pkl')
```
3. For custom plotting, yes I think Trains handle it already. I saw this example for FastAI https://allegro.ai/docs/examples/frameworks/fastai/fastai_with_tensorboard/ seems automatically log train/val loss, could it be done with lightgbm callback?

**Comment by bmartinn** [comment]
Thanks @noklam 
1. We will need to add auto-magic support for these specific calls, but totally doable.
2. If you are using joblib, *Trains* will automatically log these models (loading and storing). This means that if you are passing `output_uri` in [Task.init](https://allegro.ai/docs/task.html#trains.task.Task.init)  it will also copy them automatically to your central storage (S3 GS shared folder etc.)
3. You mean when your callback reports to TB ? If the anser is yes, then it will work out of the box (as you noted)

**Comment by noklam** [comment]
@bmartinn 
I mean callback of lightgbm, I think recording train/validation loss is always needed.

https://lightgbm.readthedocs.io/en/latest/pythonapi/lightgbm.record_evaluation.html

**Comment by bmartinn** [comment]
Hi @noklam , 
Are you suggesting we log everything in the `eval_result` dict in the [record_evaluation](https://lightgbm.readthedocs.io/en/latest/pythonapi/lightgbm.record_evaluation.html) callback automatically ?
(I'm not saying it is not doable, I'm just trying to make sure I understand you correctly)

**Comment by noklam** [comment]
I mean if I used `record_evaluation` callback, or `trains` could create a similar callback to be imported directly.

For example, mlflow did a similar thing that they will log the callback into the metrics automatically.

https://www.mlflow.org/docs/latest/_modules/mlflow/lightgbm.html#autolog:~:text=%40gorilla.patch(lightgbm),%22%22%22

**Comment by bmartinn** [comment]
@noklam yes, this seems like automagic logging of the callback dictionary.
Okay I think we have enough for a full lightgbm auto-logging support :)
I'll update here once an RC is ready :)

**Comment by allegroai-git** [comment]
Closing this as LightGBM support was added as of ClearML/Trains v0.16.3 🙂
